### PR TITLE
GitHub Issue NOAA-EMC/GSI#120. merge GFS v16.1.1 DA changes into master

### PR DESCRIPTION
**Pull request by RussTreadon-NOAA**
This PR correctly merges GFS v16.1.1 DA changes from a forked snapshot of release/gfsda.v16.1.0_merge into the authoritative master. The changes in release/gfsda.v16.1.0_merge are documented in issue #120. Additional details are found in [SCN21-52](https://www.weather.gov/media/notification/pdf2/scn21-52_gfsv16.1.1.pdf).

### CODE REVIEW

**Review by MichaelLueken-NOAA — COMMENTED** [review]
Hi @RussTreadon-NOAA,

I have completed my review of these changes.  There are a handful of minor issues with src/gsi/setupbend.f90 (replacing .le. and .gt. with symbols, adding proper precision to real values, and addressing an issue with indentation) and src/enkf/gridio_gfs.f90 (removal of commented out debug prints).  The regression tests were ran and 

- global_C96_fv3aerorad_regression_results.txt
- global_fv3_4denvar_C192_regression_results.txt
- global_fv3_4denvar_T126_regression_results.txt

Failed due to reproducibility between NOAA-EMC/GSI master and the changes in this PR.  I didn't see any reproducibility issues being indicated in issue #120.  Do you see similar reproducibility failures?

The code was compiled and ran in debug mode and the code compiled without issue and the debug tests ran through to completion.

Please address the issues in the two files and double check to see if you are seeing reproducibility issues for the three configurations.

Thanks,

Mike

**Comment by RussTreadon-NOAA** [comment]
Thanks, Mike, for the quick review.  Let me work through the corrections.
I will inform the developers who added these changes so that they are aware
of these coding standards.   I have not yet run the full suite of
regression tests.  Let me do so.

When all changes are made in my forked working copy of the master, I need
to be careful and ensure that the resulting commit only appears as 1 commit
in the PR.  The GSI wiki contains the section "How to rebase the master
into your active PR".   This applies to bringing updates from the
authoritative master into a developer PR.  Do we have comparable
instructions for how to commit reviewer requested revisions into an active
PR?

**Comment by MichaelLueken-NOAA** [comment]
Hi @RussTreadon-NOAA,

For updating the PR to ensure that there is only a single commit, please see Redmine's:

https://vlab.noaa.gov/redmine/projects/comgsi/wiki/How_to_make_modifications_once_you_have_pushed_your_changes_to_code_review

git commit --amend

will allow you to commit changes and not make additional commits.  The only difference is that you will need to use:

git push origin master --force

to push changes back to repository.

Mike

**Comment by RussTreadon-NOAA** [comment]
**WCOSS_D Regression Tests**

As Mike found, the following regression tests fail on WCOSS_D:
- `global_C96_fv3aerorad`
- `global_fv3_4denvar_T126`
- `global_fv3_4denvar_C192`
 
Failure of the `global_fv3_4denvar_C192` test is correct.   This test writes netcdf analysis increment files.  `compare_ncfile.py` of the `updat` and `contrl` `siginc.nc` files show differences in the `delz` increment.   This is expected.  PR #169 corrects an error in the sign of the delz increment in `src/gsi/netcdfgfs_io.f90`.   This bug fix in not in the authoritative repo master.  

Failures of the `global_C96_fv3aerorad` and `global_fv3_4denvar_T126` tests are correct.  Both cases use fv3 backgrounds in nemsio format.   `nemsio_read siganl` shows variable `delz` differs between the `updat` and `contrl` `global_gsi.x`  This is expected.  PR #169 corrects an error in the sign of the delz increment in `src/gsi/ncepnems_io.f90`.   This bug fix is not in the authoritative repo master.

As a test, the PR #169 change to `src/gsi/ncepnems_io.f90` was reverted in a local working copy, `global_gsi.x` rebuilt, and the `global_fv3_4denvar_T126` rerun.   The `global_fv3_4denvar_T126` test passed.   The resulting `siganl` files are b4b identical with those generated by the authoritative repo master.

For additional information regarding the delz bug files, please see issue #117 and PR #116.

File: src/enkf/gridio_gfs.f90 [hunk 1]
```diff
@@ -1509,16 +1513,19 @@ subroutine writegriddata_pnc(vars3d,vars2d,n3d,n2d,levels,ndim,grdin,no_inflate_
...
+        ! (note that ak,bk are already reversed to go from bottom to top)
+        ug=ug*log((100_r_kind*ak(krev+1)+bk(krev+1)*vg)/(100_r_kind*ak(krev)+bk(krev)*vg))
         ! ug is hydrostatic analysis delz inferred from analysis ps,Tv
-        ! delzb is hydrostatic background delz inferred from background ps,Tv
+        ! delzb is (negative) hydrostatic background delz inferred from background ps,Tv
         delzb=(rd/grav)*reshape(tv_bg(:,:,ki),(/nlons*nlats/))
-        delzb=delzb*log((100_r_kind*ak(krev)+bk(krev)*values_1d)/(100_r_kind*ak(krev+1)+bk(krev+1)*values_1d))
-        ug3d(:,:,ki)=values_3d(:,:,ki) + reshape(delzb-ug,(/nlons,nlats/))
+        delzb=delzb*log((100_r_kind*ak(krev+1)+bk(krev+1)*values_1d)/(100_r_kind*ak(krev)+bk(krev)*values_1d))
+        !print *,'writegriddata_pnc: min/max delzb',k,minval(delzb),maxval(delzb)
```
[the remark this answers is not in the document]
Comment (RussTreadon-NOAA) [reply src/enkf/gridio_gfs.f90]: Debug print removed at f027f58.

File: src/enkf/gridio_gfs.f90 [hunk 2]
```diff
@@ -2475,8 +2484,10 @@ subroutine writegriddata(nanal1,nanal2,vars3d,vars2d,n3d,n2d,levels,ndim,grdin,n
         if (hasfield) then
            call nemsio_readrecv(gfilein,'pres','sfc',1,nems_wrk2,iret=iret)
            delzb=(rd/grav)*nems_wrk
-           ! ps in Pa here, need to multiply ak by 100.
-           delzb=delzb*log((100_r_kind*ak(k)+bk(k)*nems_wrk2)/(100_r_kind*ak(k+1)+bk(k+1)*nems_wrk2))
+           ! ps in Pa here, need to multiply ak by 100. k is bottom to top,
+           ! calculate delzb so it is negative.
+           delzb=delzb*log((100_r_kind*ak(k+1)+bk(k+1)*nems_wrk2)/(100_r_kind*ak(k)+bk(k)*nems_wrk2))
+           !print *,'writegriddata nemsio: min/max delzb',k,minval(delzb),maxval(delzb)
```
[the remark this answers is not in the document]
Comment (RussTreadon-NOAA) [reply src/enkf/gridio_gfs.f90]: Debug print removed at f027f58.

File: src/enkf/gridio_gfs.f90 [hunk 3]
```diff
@@ -2876,17 +2889,21 @@ subroutine writegriddata(nanal1,nanal2,vars3d,vars2d,n3d,n2d,levels,ndim,grdin,n
...
            ! ug is hydrostatic analysis delz inferred from analysis ps,Tv
            ! delzb is hydrostatic background delz inferred from background ps,Tv
-           delzb=(rd/grav)*reshape(tv_bg(:,:,nlevs-k+1),(/nlons*nlats/))
-           delzb=delzb*log((100_r_kind*ak(k)+bk(k)*values_1d)/(100_r_kind*ak(k+1)+bk(k+1)*values_1d))
-           ug3d(:,:,nlevs-k+1)=values_3d(:,:,nlevs-k+1) +&
-           reshape(delzb-ug,(/nlons,nlats/))
+           ! calculate so it is negative
+           delzb=(rd/grav)*reshape(tv_bg(:,:,k),(/nlons*nlats/))
+           delzb=delzb*log((100_r_kind*ak(krev+1)+bk(krev+1)*values_1d)/(100_r_kind*ak(krev)+bk(krev)*values_1d))
+           !print *,'writegriddata netcdf: min/max delzb',k,minval(delzb),maxval(delzb)
```
[the remark this answers is not in the document]
Comment (RussTreadon-NOAA) [reply src/enkf/gridio_gfs.f90]: Debug print removed at f027f58.

File: src/enkf/gridio_gfs.f90 [hunk 4]
```diff
@@ -3571,15 +3590,18 @@ subroutine writeincrement(nanal1,nanal2,vars3d,vars2d,n3d,n2d,levels,ndim,grdin,
...
+        ! calculate ug (analysis delz) so it is negative. 
+        ! (note that ak,bk are already reversed to go from bottom to top)
+        ug=ug*log((100_r_kind*ak(krev+1)+bk(krev+1)*vg)/(100_r_kind*ak(krev)+bk(krev)*vg))
         ! ug is hydrostatic analysis delz inferred from analysis ps,Tv
         ! delzb is hydrostatic background delz inferred from background ps,Tv
         delzb=(rd/grav)*reshape(tv(:,:,k),(/nlons*nlats/))
-        delzb=delzb*log((100_r_kind*ak(krev)+bk(krev)*psges)/(100_r_kind*ak(krev+1)+bk(krev+1)*psges))
-        inc3d(:,:,k)=reshape(delzb-ug,(/nlons,nlats/))
+        delzb=delzb*log((100_r_kind*ak(krev+1)+bk(krev+1)*psges)/(100_r_kind*ak(krev)+bk(krev)*psges))
+        !print *,'writeincrement: min/max delzb',k,minval(delzb),maxval(delzb)
```
[the remark this answers is not in the document]
Comment (RussTreadon-NOAA) [reply src/enkf/gridio_gfs.f90]: Debug print removed at f027f58.

File: src/enkf/gridio_gfs.f90 [hunk 5]
```diff
@@ -4016,12 +4040,16 @@ subroutine writeincrement_pnc(vars3d,vars2d,n3d,n2d,levels,ndim,grdin,no_inflate
...
+        ! (note that ak,bk are already reversed to go from bottom to top)
+        ug=ug*log((100_r_kind*ak(krev+1)+bk(krev+1)*vg)/(100_r_kind*ak(krev)+bk(krev)*vg))
         ! ug is hydrostatic analysis delz inferred from analysis ps,Tv
         ! delzb is hydrostatic background delz inferred from background ps,Tv
+        ! calculate delzb so it is negative
         delzb=(rd/grav)*reshape(tv(:,:,ki),(/nlons*nlats/))
-        delzb=delzb*log((100_r_kind*ak(krev)+bk(krev)*psges)/(100_r_kind*ak(krev+1)+bk(krev+1)*psges))
-        inc3d(:,:,ki)=reshape(delzb-ug,(/nlons,nlats/))
+        delzb=delzb*log((100_r_kind*ak(krev+1)+bk(krev+1)*psges)/(100_r_kind*ak(krev)+bk(krev)*psges))
+        !print *,'writeincrement_pnc: min/max delzb',k,minval(delzb),maxval(delzb)
```
[the remark this answers is not in the document]
Comment (RussTreadon-NOAA) [reply src/enkf/gridio_gfs.f90]: Debug print removed at f027f58.

File: src/gsi/setupbend.f90 [hunk 6]
```diff
@@ -579,290 +580,295 @@ subroutine setupbend(obsLL,odiagLL, &
...
+             elseif (tpdpres(i) < ref_rad(top_layer_SR+1)) then !obs below model close-to-SR layer
+                qcfail(i)=.true.
+             elseif (tpdpres(i) >= ref_rad(top_layer_SR+1) .and. tpdpres(i) <= ref_rad(top_layer_SR+2)) then !source too close
+                    qcfail(i)=.true.
+             else !above
+             endif
+          endif
+
+!         check for SR in obs, will be updated in genstats. 
+          if ( data(igps,i) >= 0.03 .and. qc_layer_SR) then
```
[the remark this answers is not in the document]
Comment (RussTreadon-NOAA) [reply src/gsi/setupbend.f90]: `0.03` replaced by `0.03_r_kind` at f027f58.

File: src/gsi/setupbend.f90 [hunk 7]
```diff
@@ -579,290 +580,295 @@ subroutine setupbend(obsLL,odiagLL, &
...
+                repe_gps=-1.87788_r_kind+0.354718_r_kind*alt-0.00313189_r_kind*alt**2
+              else
+                repe_gps=-2.41024_r_kind+0.806594_r_kind*alt-0.027257_r_kind*alt**2
+              endif
+           endif
+         else 
+!        CDAAC-type processing
+           if ((data(isatid,i) > 749).and.(data(isatid,i) < 756)) then
+              if ((data(ilate,i)> r40).or.(data(ilate,i)< -r40)) then
+                if (alt.le.8.0_r_kind) then
```
[the remark this answers is not in the document]
Comment (RussTreadon-NOAA) [reply src/gsi/setupbend.f90]: `.le.` replaced by `<=` at f027f58.

File: src/gsi/setupbend.f90 [hunk 8]
```diff
@@ -579,290 +580,295 @@ subroutine setupbend(obsLL,odiagLL, &
...
+                  repe_gps=-1.0304261_r_kind+0.3203316_r_kind*alt+0.0141337_r_kind*alt**2
+                elseif (alt.gt.8.0_r_kind.and.alt.le.r12) then
```
[the remark this answers is not in the document]
Comment (RussTreadon-NOAA) [reply src/gsi/setupbend.f90]: `.gt.` replaced by `>`.  `.le.` replaced by `<=` at f027f58.

File: src/gsi/setupbend.f90 [hunk 9]
```diff
@@ -579,290 +580,295 @@ subroutine setupbend(obsLL,odiagLL, &
...
+                  repe_gps=2.1750271_r_kind+0.0431177_r_kind*alt-0.0008567_r_kind*alt**2
+                else
+                  repe_gps=-0.3447429_r_kind+0.2829981_r_kind*alt-0.0028545_r_kind*alt**2
+                endif
+              else
+                if (alt.le.4.0_r_kind) then
```
[the remark this answers is not in the document]
Comment (RussTreadon-NOAA) [reply src/gsi/setupbend.f90]: `.le.` replaced by `<=` at f027f58.

File: src/gsi/setupbend.f90 [hunk 10]
```diff
@@ -579,290 +580,295 @@ subroutine setupbend(obsLL,odiagLL, &
...
+                  repe_gps=0.7285212_r_kind-1.1138755_r_kind*alt+0.2311123_r_kind*alt**2
+                elseif (alt.le.r18.and.alt.gt.4.0_r_kind) then
```
[the remark this answers is not in the document]
Comment (RussTreadon-NOAA) [reply src/gsi/setupbend.f90]: `.le.` replaced by `<=`.  `.gt.` replaced by `>` at f027f58.

File: src/gsi/setupbend.f90 [hunk 11]
```diff
@@ -1237,7 +1243,7 @@ subroutine setupbend(obsLL,odiagLL, &
         gps_alltail(ibin)%head%obserr   = data(ier,i)
         gps_alltail(ibin)%head%dataerr  = data(ier,i)*data(igps,i)
         gps_alltail(ibin)%head%muse     = muse(i) ! logical
-     endif ! (last_pass)
+  endif ! (last_pass)
   end do ! i=1,nobs
```
[the remark this answers is not in the document]
Comment (RussTreadon-NOAA) [reply src/gsi/setupbend.f90]: Indentation between lines 981 and 1246 corrected at f027f58.

File: src/gsi/setupbend.f90 [hunk 12]
```diff
@@ -579,290 +580,295 @@ subroutine setupbend(obsLL,odiagLL, &
...
-!       get pressure (in mb), temperature and moisture at obs location
-        call tintrp31(ges_lnprsi(:,:,1:nsig,:),dpressure,dlat,dlon,hob,&
+       if (alt <= six) then
+          if (top_layer_SR >= 1) then ! SR exists for at least one layer. Check if obs is inside 
+             if (tpdpres(i)==ref_rad(top_layer_SR+1)) then !obs inside model SR layer
+                qcfail(i)=.true.
+             elseif (tpdpres(i) < ref_rad(top_layer_SR+1)) then !obs below model close-to-SR layer
+                qcfail(i)=.true.
+             elseif (tpdpres(i) >= ref_rad(top_layer_SR+1) .and. tpdpres(i) <= ref_rad(top_layer_SR+2)) then !source too close
+                    qcfail(i)=.true.
```
Comment (MichaelLueken-NOAA) [inline src/gsi/setupbend.f90:590]: Please remove the extra indentation on line 590.

File: src/gsi/setupbend.f90 [hunk 13]
```diff
@@ -579,290 +580,295 @@ subroutine setupbend(obsLL,odiagLL, &
...
+             else !above
```
Comment (MichaelLueken-NOAA) [inline src/gsi/setupbend.f90:591]: Will modifications be made for obs above model?  If not, should line 591 be removed?

**Review by MichaelLueken-NOAA — COMMENTED** [review]
One final comment, then this work will be ready for the review committee.  On line 590, please remove the extra indentation (qcfail should align with the previous entries).  Additionally, will there be anything for the above section of the if statement logic (line 591).  If not, then should this line be removed as well?

File: src/gsi/setupbend.f90 (hunk 12 above)
Comment (RussTreadon-NOAA) [reply src/gsi/setupbend.f90]: Indentation removed on line 590 at   d199ed5.    

Add print statement in the empty else branch of `src/gsi/setupbend.f90`.  Build `global_gsi.x` and run executable for four gfs cases:  2021053112 to 2021060106.   Each case generated printout from the empty else branch, on average 13621 entries per outer loop.   Removed the else branch from `src/gsi/setupbend.f90`.  Build `global_gsi.x` and reran the same four gfs cases.  The resulting analyses are identical.    This is not surprising.   The empty else branch currently does nothing.   Removing the empty else branch should not alter results.   

Two questions remain
1. Why is the else statement in the code?
2. What code, if any, should be in the empty else branch

Kristen asked Hui these questions.  Both Kristen and Hui agree the empty else branch may be removed.   

Remove empty else branch at d199ed5.

File: src/gsi/setupbend.f90 (hunk 13 above)
Comment (RussTreadon-NOAA) [reply src/gsi/setupbend.f90]: Based on replies from Kristen and Hui, the empty else block should be removed.  As such, the empty else block was removed from `src/gsi/setupbend.f90` at d199ed5.

**Comment by MichaelLueken-NOAA** [comment]
The due date for the review committee has past with no feedback.  I will now give final approval to these changes and merge them to the authoritative repository.

**Pull request merged**